### PR TITLE
Add LLVM_Util methods to detect appropriate ISA for the host CPU.

### PR DESCRIPTION
**Pull request by lgritz**
We don't use this quite yet, but are staging it for easy review.

Work by Alex Wells. Some refactoring and minor edits by LG.

Signed-off-by: Larry Gritz <lg@larrygritz.com>

### CODE REVIEW

File: src/liboslexec/llvm_util.cpp [hunk 1]
```diff
@@ -437,7 +437,369 @@ LLVM_Util::end_builder ()
...
+    return target_isa_names[static_cast<int>(isa)];
+}
+
+
+
+bool
+LLVM_Util::detect_cpu_features(string_view requested_isa_string,
+                               bool no_fma)
+{
+    TargetISA requestedISA = lookup_isa_by_name(requested_isa_string);
```
Comment (AlexMWells) [inline src/liboslexec/llvm_util.cpp:641]: I am wondering if maybe `requested_isa_string` parameter to detect_cpu_features, should just be `requestedISA `and make caller call` lookup_isa_by_name(requested_isa_string) `themselves?

File: src/liboslexec/llvm_util.cpp [hunk 2]
```diff
@@ -437,7 +437,369 @@ LLVM_Util::end_builder ()
...
+    "pclmul",
+    // "pku" Memory Protection Keys we shouldn't need/require this feature,
+    "popcnt", "prfchw", "rdrnd", "rdseed", "rtm",
+#if OSL_LLVM_VERSION >= 70
+    "sahf",
+#endif
+    "sse", "sse2", "sse3", "sse4.1", "sse4.2", "ssse3", "x87"
+    // , "xsave", "xsavec", "xsaveopt", "xsaves" // Save Processor Extended States, we don't use
+};
+
```
Comment (AlexMWells) [inline src/liboslexec/llvm_util.cpp:563]: We have discovered the list of cpu features is a bit too specific to server parts and could have issues with client parts that may not have some of the cpu features, like "rtm".  As it turns out most of these server/OS geared instructions are only really accessible explicitly with intrinsics and we don't expect LLVM, GCC, CLang, or ICC to generate them by themselves, so we should be safe to remove those features which we are not explicitly using.  Suggest changing these lists to:
```
// clang: default
// icc: default
static const char * required_cpu_features_by_x64[] = {
    "fxsr", "mmx", "sse", "sse2", "x87"
};

// clang: -march=nehalem
// icc: -xSSE4.2
static const char * required_cpu_features_by_SSE4_2[] = {
    "cx16","fxsr","mmx","popcnt",
#if OSL_LLVM_VERSION >= 70
    // "sahf", // we shouldn't need/require this feature
#endif
    "sse","sse2","sse3","sse4.1",
    "sse4.2","ssse3", "x87"
};

// clang: -march=corei7-avx
// icc: -xAVX
static const char * required_cpu_features_by_AVX[] = {
    "aes", "avx", "cx16", "fxsr", "mmx", "pclmul", "popcnt",
#if OSL_LLVM_VERSION >= 70
    // "sahf", // we shouldn't need/require this feature
#endif
    "sse",
    "sse2", "sse3", "sse4.1", "sse4.2", "ssse3", "x87"
    // ,"xsave","xsaveopt" // Save Processor Extended States, we don't use
};

// clang: -march=core-avx2
// icc: -xCORE-AVX2
static const char * required_cpu_features_by_AVX2[] = {
    "aes", "avx", "avx2", "bmi", "bmi2", "cx16", "f16c", "fma",
    // "fsgsbase", // we shouldn't need/require this feature
    "fxsr",
    // "invpcid", // Invalidate Process-Context Identifier, we don't use
    "lzcnt", "mmx", "movbe", "pclmul", "popcnt",
    // "rdrnd", // random # don't require unless we make use of it
#if OSL_LLVM_VERSION >= 70
    // "sahf", // we shouldn't need/require this feature
#endif
    "sse",
    "sse2", "sse3", "sse4.1", "sse4.2", "ssse3", "x87"
    // ,"xsave","xsaveopt" // // Save Processor Extended States, we don't use
};

// clang: -march=core-avx2 -mno-fma
// icc: -xCORE-AVX2 -no-fma
static const char * required_cpu_features_by_AVX2_noFMA[] = {
    "aes", "avx", "avx2", "bmi", "bmi2", "cx16", "f16c",
    // "fsgsbase", // we shouldn't need/require this feature
    "fxsr",
    // "invpcid", // Invalidate Process-Context Identifier, we don't use
    "lzcnt", "mmx", "movbe", "pclmul", "popcnt",
    // "rdrnd", // random # don't require unless we make use of it
#if OSL_LLVM_VERSION >= 70
    // "sahf", // we shouldn't need/require this feature
#endif
    "sse",
    "sse2", "sse3", "sse4.1", "sse4.2", "ssse3", "x87"
    // , "xsave", "xsaveopt" // Save Processor Extended States, we don't use
};

// clang: -march=skylake-avx512
// icc: -xCORE-AVX512
static const char * required_cpu_features_by_AVX512[] = {
    "adx", "aes", "avx", "avx2", "avx512bw", "avx512cd", "avx512dq",
    "avx512f", "avx512vl", "bmi", "bmi2",
    // "clflushopt", "clwb", flushing for volatile/persistent memory we shouldn't need
    "cx16",
    "f16c", "fma",
    // "fsgsbase", // we shouldn't need/require this feature,
    "fxsr",
    // "invpcid", // Invalidate Process-Context Identifier, we don't use
    "lzcnt", "mmx", "movbe",
    //"mpx", // Memory Protection Extensions, we don't use
    "pclmul",
    // "pku"//  Memory Protection Keys we shouldn't need/require this feature,
    "popcnt",
    // "prfchw", // prefetch wide we shouldn't need/require this feature,
    // "rdrnd", "rdseed", // random # don't require unless we make use of it
    // "rtm", // transaction memory we shouldn't need/require this feature,
#if OSL_LLVM_VERSION >= 70
    // "sahf", // we shouldn't need/require this feature
#endif
    "sse", "sse2", "sse3", "sse4.1", "sse4.2", "ssse3", "x87"
    // , "xsave", "xsavec", "xsaveopt", "xsaves" // Save Processor Extended States, we don't use
};
// clang: -march=skylake-avx512 -mno-fma
// icc: -xCORE-AVX512 -no-fma
static const char * required_cpu_features_by_AVX512_noFMA[] = {
    "adx", "aes", "avx", "avx2", "avx512bw", "avx512cd", "avx512dq",
    "avx512f", "avx512vl", "bmi", "bmi2",
    // "clflushopt", "clwb", flushing for volatile/persistent memory we shouldn't need
    "cx16",
    "f16c",
    // "fsgsbase", // we shouldn't need/require this feature
    "fxsr",
    // "invpcid", // Invalidate Process-Context Identifier, we don't use
    "lzcnt", "mmx", "movbe",
    //"mpx", // Memory Protection Extensions, we don't use
    "pclmul",
    // "pku"//  Memory Protection Keys we shouldn't need/require this feature,
    "popcnt",
    // "prfchw", // prefetch wide we shouldn't need/require this feature,
    // "rdrnd", "rdseed", // random # don't require unless we make use of it
    // "rtm", // transaction memory we shouldn't need/require this feature,
#if OSL_LLVM_VERSION >= 70
    // "sahf", // we shouldn't need/require this feature
#endif
    "sse", "sse2", "sse3", "sse4.1", "sse4.2", "ssse3", "x87"
    // , "xsave", "xsavec", "xsaveopt", "xsaves" // Save Processor Extended States, we don't use
};
```

Also if our minimum LLVM version is 7+ now, we could drop the conditional compilation pre-processor stuff:
```
#if OSL_LLVM_VERSION >= 70
```

File: src/liboslexec/llvm_util.cpp (hunk 1 above)
Comment (lgritz) [reply src/liboslexec/llvm_util.cpp:641]: I have no preference. Would you like me to change it?

File: src/liboslexec/llvm_util.cpp (hunk 2 above)
Comment (lgritz) [reply src/liboslexec/llvm_util.cpp:582]: Thanks. Pushing an amended commit now with this change.

File: src/liboslexec/llvm_util.cpp (hunk 1 above)
Comment (AlexMWells) [reply src/liboslexec/llvm_util.cpp:659]: Yes, I think so, it was bugging me in use, and now that you have busted out detect_cpu_features to its own function, it makes sense.

Comment (lgritz) [reply src/liboslexec/llvm_util.cpp:659]: OK, will do.

Comment (lgritz) [reply src/liboslexec/llvm_util.cpp]: Amended.

Also make llvmutil_test double check that what it's getting from these LLVM checks of the host ISA matches what OIIO thinks it's getting from cpuid queries.

**Comment by AlexMWells** [comment]
LLVM_Util::LLVM_Util() constructor should probably set
m_target_isa(TargetISA::UNKNOWN)

**Pull request merged**